service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/C21OrderController.java

```diff
@@ -0,0 +1,127 @@
+            .build();
+    }
+
+    @PostMapping("/mid-event")
+    public AboutToStartOrSubmitCallbackResponse handleMidEvent(
+        @RequestHeader(value = "authorization") String authorization,
+        @RequestHeader(value = "user-id") String userId,
+        @RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        Map<String, Object> data = caseDetails.getData();
+        CaseData caseData = mapper.convertValue(data, CaseData.class);
+        String index = generateIndexForFileName(caseData.getC21OrderBundle());
+
+        Document c21Document = getDocument(
+            authorization,
+            userId,
+            index,
+            createC21OrderService.getC21OrderTemplateData(caseData));
+
+        data.put("temporaryC21Order", addDocumentToC21Order(caseData, c21Document));
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(data).build();
+    }
+
+    @PostMapping("/about-to-submit")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(
+        @RequestBody CallbackRequest callbackRequest) {
+        Map<String, Object> data = callbackRequest.getCaseDetails().getData();
+        CaseData caseData = mapper.convertValue(data, CaseData.class);
+
+        data.put("c21OrderBundle", createC21OrderService.addToC21OrderBundle(
+            caseData.getTemporaryC21Order(), caseData.getJudgeAndLegalAdvisor(), caseData.getC21OrderBundle()));
+        data.remove("temporaryC21Order");
+        data.remove("judgeAndLegalAdvisor");
+
+        return AboutToStartOrSubmitCallbackResponse.builder().data(data).build();
+    }
+
+    private String generateIndexForFileName(List<Element<C21OrderBundle>> c21OrderBundle) {
+        return (c21OrderBundle != null) ? Integer.toString(c21OrderBundle.size() + 1) : "1";
+    }
+
+    private C21Order addDocumentToC21Order(CaseData caseData, Document document) {
+        return caseData.getTemporaryC21Order().toBuilder()
+            .c21OrderDocument(DocumentReference.builder()
+                .url(document.links.self.href)
+                .binaryUrl(document.links.binary.href)
+                .filename(document.originalDocumentName)
+                .build())
+            .orderTitle(defaultIfBlank(caseData.getTemporaryC21Order().getOrderTitle(), "Order"))
+            .build();
+    }
+
+    private Document getDocument(@RequestHeader("authorization") String authorization,
+                                 @RequestHeader("user-id") String userId,
+                                 String index,
+                                 Map<String, Object> templateData) {
+        DocmosisDocument document = docmosisService.generateDocmosisDocument(templateData, C21);
+        return uploadDocumentService.uploadPDF(userId, authorization, document.getBytes(),
+            C21.getDocumentTitle() + "_" + index + ".pdf");
+    }
+}
```

Comment:
I believe this is called a service... be nice to see some unit tests 😅

Reply:
are you saying these should be in the service + have some unit tests for them? or they can stay here, but I should write some unit tests for them (in ControllerTest? in ServiceTest?) Honestly I don't fully understand our tests and probably should have asked more questions at an earlier stage, a lot of the tests was me referencing existing tests and trying to work it out from that - to be honest it would be really helpful for me to just sit down with someone and go through the practices we follow for our unit/integration tests